gui-whelktool/src/main/resources/templates/change-complex-subject.groovy

```diff
@@ -0,0 +1,77 @@
+PrintWriter failedBibIDs = getReportWriter("failed-to-update-bibIDs")
+PrintWriter scheduledForUpdate = getReportWriter("scheduled-for-update")
+PrintWriter manualReviewLog = getReportWriter("needs-manual-review")
+
+selectBySqlWhere( "collection = 'bib' AND (" +
+        "data#>'{@graph,1,instanceOf,subject}' @> '[{\"@type\":\"ComplexSubject\"}]'::jsonb OR " +
+        "data#>'{@graph,1,subject}' @> '[{\"@type\":\"ComplexSubject\"}]'::jsonb" +
+        ")",
+        silent: false, { bib ->
+    boolean changed = false
+
+    if (bib.graph[1].instanceOf != null &&
+            bib.graph[1].instanceOf.subject != null &&
+            bib.graph[1].instanceOf.subject instanceof List) {
+        for (Object subject : bib.graph[1].instanceOf.subject) {
+            changed |= fixSubject(subject, manualReviewLog, bib.doc.shortId)
+        }
+    }
+
+    if (bib.graph[1].subject != null && bib.graph[1].subject instanceof List) {
+        for (Object subject : bib.graph[1].subject) {
+            changed |= fixSubject(subject, manualReviewLog, bib.doc.shortId)
+        }
+    }
+
+    if (changed) {
+        scheduledForUpdate.println("${bib.doc.getURI()}")
+        bib.scheduleSave(loud: true, onError: { e ->
+            failedBibIDs.println("Failed to update ${bib.doc.shortId} due to: $e")
+        })
+    }
+})
+
+boolean fixSubject(Object subject, PrintWriter manualReviewLog, String id) {
+    boolean fixedSameAs = false
+    boolean fixedPrefLabel = false
+    boolean fixedTermComponentList = false
+
+    if (subject["@type"] == "ComplexSubject") {
+
+        if (subject.sameAs != null && subject.sameAs instanceof List) {
+            for (Object ref : subject.sameAs) {
+                String encodedFrom = URLEncoder.encode("FROMTERM", "UTF-8")
+                String encodedTo = URLEncoder.encode("TOTERM", "UTF-8")
+                if (ref["@id"].contains(encodedFrom)) {
+                    ref["@id"] = ref["@id"].replaceFirst(encodedFrom, encodedTo)
+                    fixedSameAs = true
+                }
+            }
+        }
+
+        if (subject.prefLabel != null && subject.prefLabel instanceof String && subject.prefLabel.startsWith("FROMTERM")) {
+            subject.prefLabel = subject.prefLabel.replaceFirst("FROMTERM", "TOTERM")
+            fixedPrefLabel = true
+        }
+
+        if (subject.termComponentList != null && subject.termComponentList instanceof List) {
+            if (subject.termComponentList.size() > 0) {
+                if (subject.termComponentList[0]["@type"] == "Topic"
+                        && subject.termComponentList[0]["prefLabel"] == "FROMTERM") {
```

Comment:
Instead of changing the prefLabel of the local entity of the termComponentList it would be better to replace the local entity with a link to the actual subject ("TOTERM"?). We want to get rid of as many local entities as possible in ComplexSubjects.

Reply:
I really think this needs to be a separate thing we do. The task here is only to update the term itself.

Reply:
ok, fair 👍